server/interpreter.js

```diff
@@ -451,6 +449,46 @@ Interpreter.prototype.stop = function() {
   this.status = Interpreter.Status.STOPPED;
 };
 
+/**
+ * Convert source code into a ready-to-execute parse tree.
+ * @private
+ * @param {string} src The source code to be compiled.
+ * @param {!Interpreter.Owner=} perms Re-throw Acorn parse errors as
+ *     user errors owned by perms.  (Default: re-throw Acorn parse
+ *     errors as internal (native) errors.)
+ * @return {!Interpreter.Node} node Root AST node.
+ */
+Interpreter.prototype.compile_ = function(src, perms) {
+  try {
+    var ast = acorn.parse(src, Interpreter.PARSE_OPTIONS);
+  } catch (e) {  // Acorn threw a SyntaxError.  Rethrow as a trappable error?
+    throw perms ? this.errorNativeToPseudo(e, perms) : e;
+  }
+
+  (function analyse(node) {
```

Comment:
Is there some reason this isn't a private helper function?  As it stands, this function is being recreated over and over, every time compile_ is called.

Reply:
Two reasons:
* I don't expect it to be called from anywhere else (and this way it is *actually* private), and
* I foresee future changes for which having access to local variables in `.compile_` will be useful.  In particular, much of the tree-walking work presently done by `.populateScope_` can actually be done upfront in `.compile_`—so it only needs to be done once, instead of once per invocation of a function—and for that it will be convenient to have access to an array to accumulate `var` bindings and so on.